db/commitlog/commitlog_entry.cc

```diff
@@ -10,34 +10,124 @@
-        if (_with_schema) {
-            return std::move(wr).write_mapping(_schema->get_column_mapping());
-        } else {
-            return std::move(wr).skip_mapping();
+    std::visit(overloaded_functor {
+        [&](const frozen_mutation* m) {
+            [this, wr = ser::writer_of_commitlog_entry<Output>(out)] () mutable {
+                if (_with_schema) {
+                    return std::move(wr).write_mapping(_schema->get_column_mapping());
+                } else {
+                    return std::move(wr).skip_mapping();
+                }
+            }().write_mutation(*m).end_commitlog_entry();
+        },
+        [&](const mutation* m) {
+            throw std::runtime_error("Should not reach");
+        }
+    }, _mutation);
+}
+
+void commitlog_entry_writer::set_with_schema(bool value) {
+    if (std::exchange(_with_schema, value) != value || _size == std::numeric_limits<size_t>::max()) {
+        os.clear();
+        compute_size();
+    }
+}
+
+size_t commitlog_entry_writer::mutation_size() const {
+    if (_size != std::numeric_limits<size_t>::max()) {
+        return _size;
+    }
+    return std::visit(overloaded_functor {
+        [&](const frozen_mutation* m) -> size_t {
+            return m->representation().size();
+        },
+        [&](const mutation* m) -> size_t {
+            return frozen_size(*m);
         }
-    }().write_mutation(_mutation).end_commitlog_entry();
+    }, _mutation);
+}
+
+static bytes_ostream compress(bytes_ostream&& tmp, const compressor* comp) {
+    bytes_ostream compressed;
+    comp->compress(std::move(tmp).to_managed_bytes(), compressed);
+    compressed.reduce_chunk_count();
+    return compressed;
 }
 
 void commitlog_entry_writer::compute_size() {
-    seastar::measuring_output_stream ms;
-    serialize(ms);
-    _size = ms.size();
+    std::visit(overloaded_functor {
+        [&](const frozen_mutation*) {
+            if (_comp) {
+                bytes_ostream tmp;
+                serialize(tmp);
+                os = compress(std::move(tmp), _comp);
+                _size = os.size();
+                return;
+            }
+            seastar::measuring_output_stream ms;
+            serialize(ms);
+            _size = ms.size();
+        },
+        [&](const mutation* m) {
+            bytes_ostream tmp;
+            {
+                ser::writer_of_commitlog_entry<bytes_ostream> wr(tmp);
+                auto frame = _with_schema 
+                    ? std::move(wr).write_mapping(_schema->get_column_mapping())
+                    : std::move(wr).skip_mapping();
+                    ;
+                freeze(tmp, *m);
+            }
+            // now compress
+            if (_comp) {
+                tmp = compress(std::move(tmp), _comp);
+            }
+            tmp.reduce_chunk_count();
+            os = std::move(tmp);
+            _size = os.size();
+        },
+    }, _mutation);
```

Comment:
I'm confused about why we have all those dual paths, and how we ensure they match each other.

Wouldn't it be better to freeze the mutation if unfrozen? Perhaps it adds a memory copy if still unfrozen and not compressing? Is this the motivation?

Reply:
Yes, though we're only really limiting the longevity of temporary objects, since we need to freeze at some point. So I guess it could be removed in favour of just going frozen all the way. However, since another potential improvement down the line is having the writer use an alternative serialization I added the option to use a mutation directly as source.

Reply:
It looks like a large complication, let's drop it. What do current callers use?

Reply:
Code prior to this patch explicitly creates frozen mutation which is kept alive in scope by caller and given to this object. The idea with allowing a mutation as source and just using a temporary object here was to avoid very large mutations to exist in uncompressed frozen state for longer than required (adding a mutation to commitlog is a potentially task-switching op, so can be a relatively "long" time holding up to 200+% of the serialized size (uncompressed + compressed)

Reply:
@avikivity do you still want to drop this change?

Reply:
It _was_ dropped.